contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/oauth/AccessTokenRepository.java

```diff
@@ -0,0 +1,149 @@
+
+  private static final Logger logger = LoggerFactory.getLogger(AccessTokenRepository.class);
+  private String accessToken;
+  private final OkHttpClient client;
+  private final HttpOAuthConfig oAuthConfig;
+  private final StoragePluginRegistry registry;
+  private final HttpStoragePluginConfig pluginConfig;
+
+  public AccessTokenRepository(HttpOAuthConfig oAuthConfig,
+                               HttpProxyConfig proxyConfig,
+                               HttpStoragePluginConfig pluginConfig,
+                               StoragePluginRegistry registry) {
+    Builder builder = new OkHttpClient.Builder();
+    this.oAuthConfig = oAuthConfig;
+    this.registry = registry;
+    this.pluginConfig = pluginConfig;
+
+    if (oAuthConfig.tokens() != null && oAuthConfig.tokens().containsKey("accessToken")) {
+      accessToken = oAuthConfig.tokens().get("accessToken");
+    }
+
+    // Add proxy info
+    SimpleHttp.addProxyInfo(builder, proxyConfig);
+    client = builder.build();
+  }
+
+  /**
+   * Returns the current access token.  Does not perform an HTTP request.
+   * @return The current access token.
+   */
+  public String getAccessToken() {
+    logger.debug("Getting Access token");
+    if (accessToken == null) {
+      return refreshAccessToken();
+    }
+    return accessToken;
+  }
+
+  /**
+   * Refreshes the access token using the code and other information from the HTTP OAuthConfig.
+   * This executes a POST request.  This method will throw exceptions if any of the required fields
+   * are empty.  This plugin also updates the configuration in the storage plugin registry.
+   *
+   * In the event that a user submits a request and the access token is expired, the API will
+   * return a 401 non-authorized response.  In the event of a 401 response, the AccessTokenAuthenticator will
+   * create additional calls to obtain an updated token. This process should be transparent to the user.
+   *
+   * @return String of the new access token.
+   */
+  public String refreshAccessToken() {
+    Request request;
+    logger.debug("Refreshing Access Token.");
+    validateKeys();
+
+    // If the refresh token is present process with that
+    if (oAuthConfig.tokens().containsKey("refreshToken") &&
+      StringUtils.isNotEmpty(oAuthConfig.tokens().get("refreshToken"))) {
+      request = OAuthUtils.getAccessTokenRequestFromRefreshToken(oAuthConfig);
+    } else {
+      request = OAuthUtils.getAccessTokenRequest(oAuthConfig);
+    }
+
+    // Update/Refresh the tokens
+    Map<String, String> tokens = OAuthUtils.getOAuthTokens(client, request);
+    HttpOAuthConfig updatedConfig = new HttpOAuthConfig(oAuthConfig, tokens);
+
+    if (tokens.containsKey("accessToken")) {
+      accessToken = tokens.get("accessToken");
+    }
+
+    // This null check is here for testing only.  In actual Drill, the registry will not be null.
+    if (registry != null) {
+      OAuthUtils.updateOAuthTokens(registry, updatedConfig, pluginConfig);
```

Comment:
Please see the previous comment regarding storing access tokens in ZK...

Reply:
This is completely refactored to store (and update) the creds in the `credentialProvider`